Add functions to ContextualMenuItem to open and close menus on command

kelseyyoung opened the pull request:
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ npm run change`

#### Description of changes

This work is related to keytips. It gives a way for a developer, given a ref to a contextual menu item, to open and close its menus on command instead of mocking a click or keypress. They are similar to the functions added to BaseButton (openMenu and dismissMenu). The functions added are:

- **openSubMenu()** - opens an item's subMenu, if it has one
- **dismissSubMenu()** - closes an item's subMenu, if it has one
- **dismissMenu(dismissAll?: boolean)** - closes the menu this ContextualMenuItem lives in, or all menus if dismissAll is true. The dismissAll parameter is dependent on the developer's implementation of onDismiss (which is how it's always been)

For ContextualMenuItem, these functions will call optional props that are assigned by ContextualMenu. The openSubMenu prop function requires a target to attach the subMenu to; this is also assigned in ContextualMenu by passing down a ref to the container element (the anchor in _renderAnchorMenuItem and the button in _renderButtonItem)

ContextualMenuSplitButton also uses this interface and has access to these functions. It does not need to take in a ref for openSubMenu because the containing div (this._splitButton) already has a ref and is what the subMenu should attach to

Part of the change requires ContextualMenuItem to become a BaseComponent to take advantage of componentRef

#### Focus areas to test

Wrote unit tests and tested manually

jspurlin commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -499,6 +500,10 @@ export class ContextualMenu extends BaseComponent<IContextualMenuProps, IContext
+    const anchorRef = createRef<HTMLAnchorElement>();
+    const openSubMenu = this._onItemSubMenuExpand.bind(this);
+    const dismissSubMenu = this._onSubMenuDismiss.bind(this);
+    const dismissMenu = this.dismiss.bind(this, undefined);
```
I don't think you need to bind these sine the functions are defined a variables the `this` should be correct. And for dismissMenu you shound't need to pass in undefined since the parameters are optional

kelseyyoung replied:
I'm passing undefined because this.dismiss takes in the event (optional) as the first param but dismissMenu takes in one param which is dismissAll, which is the second param in this.dismiss. So I'm padding the params basically

kelseyyoung replied:
I will look into not binding them. I was just following the pattern throughout the file where there are binds

kelseyyoung commented:
Hmmm I will see why I'm adding 350B...I wouldn't think it would be that much based on the changes

joschect commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -599,6 +610,10 @@ export class ContextualMenu extends BaseComponent<IContextualMenuProps, IContext
         hasMenu: true
       };
     }
+    const btnRef = createRef<HTMLButtonElement>();
```
We really should avoid using refs whenever possible. Especially not in the way that this one is used. Instead of passing a ref down, pass a queryselector through or something like that. If anything the item itself should decide what the target is.

kelseyyoung replied:
The ref is required so that the submenu positions itself correctly on the containing <button> and <a> elements (which is what they normally position on) which aren't available in ContextualMenuItem. An alternative approach is to move out the button and anchor items into their own component, like ContextualMenuSplitButton was moved. This would remove the need to pass down a ref (in the code you can see that I don't pass down that ref to ContextualMenuSplitButton). Again curious to hear @dzearing and @cliffkoh opinions

kelseyyoung replied:
Also up for discussion: passing the componentRef to ContextualMenuItem through IContextualMenuItem and 'renderItemRef' in order to set the componentRef on ContextualMenuItem. Again pinging @dzearing and @cliffkoh

jspurlin replied:
@kelseyyoung the best sounding approach would be to split button and anchor into their own components ( ContextualMenuButton and ContextualMenuAnchor) to remove the need to pass down a ref

jspurlin commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -635,6 +652,10 @@ export class ContextualMenu extends BaseComponent<IContextualMenuProps, IContext
     const { contextualMenuItemAs } = this.props;
 
+    const openSubMenu = this._onItemSubMenuExpand.bind(this);
+    const dismissSubMenu = this._onSubMenuDismiss.bind(this);
```
Do these need to use bind?

kelseyyoung commented:
@jspurlin and @joschect, I've updated the PR where Button and Anchor are moved to their own components, I've called these along with Split Button "ContextualMenuItemWrapper" components. This change means no real changes to ContextualMenuItem. The componentRef in IContextualMenuItem is now passed to one of these wrapper components

kelseyyoung commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -547,79 +519,35 @@ export class ContextualMenu extends BaseComponent<IContextualMenuProps, IContext
-    };
-
-    let { keytipProps } = item;
-    if (keytipProps && itemHasSubmenu) {
```
This was not properly copied over, will fix

joschect replied:
Have you fixed this?

kelseyyoung commented:
After testing in WAC I'm having some problems where the internal ref that determines where to open the submenu isn't being set fast enough. I'll try to figure out why tomorrow

kelseyyoung commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuAnchor.tsx:
```diff
@@ -0,0 +1,136 @@
+
+  public dismissSubMenu = (): void => {
+    const { item, dismissSubMenu } = this.props;
+    if (hasSubmenu(item) && dismissSubMenu) {
```
anchor shouldn't be able to do any of these because it doesn't have a submenu ever

kelseyyoung commented:
Another update: the previous iteration had issues where our internal refs weren't being correctly set. I've moved the open/dismiss functions back to ContextualMenuItem. For openSubMenu to get the element to position on, I've added a 'getContainerElement' prop in ContextualMenuItem which is set by the wrapper components (SplitButton, Button, Anchor). This involves refs internally in the wrapper components but doesn't involve passing them down to ContextualMenuItem. Would love to hear feedback from @joschect and @dzearing

jspurlin commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItem.tsx:
```diff
@@ -67,19 +67,46 @@ const renderSubMenuIcon = ({ item, classNames }: IContextualMenuItemProps) => {
 
-export const ContextualMenuItem: React.StatelessComponent<IContextualMenuItemProps> = (props) => {
-  const { item, classNames } = props;
+export class ContextualMenuItem extends BaseComponent<IContextualMenuItemProps, {}> {
```
Why did the extends change from React.StatelessComponent to BaseComponent?

jspurlin replied:
Any decision here?

jspurlin commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuAnchor.tsx:
```diff
@@ -0,0 +1,136 @@
+          disabled={ disabled }
+        >
+          { (keytipAttributes: any): JSX.Element => (
+            <a
```
ContextualMenuAnchor and ContextualMenuButton are very similar (a different tag is used and a few different props are used). Did you consider creating a generic component that takes a tag and itemprops and would return the correct type of item?

joschect replied:
This isn't a bad idea, you could even have the tag be a prop that get's passed down.

kelseyyoung replied:
Jeremy and I talked about it; in my opinion there are still a few differences between the two and may require more work. I don't know if this PR should be that far reaching. I understand there's now some duplication across these two files, but I guess I would rather log a bug to fix this later

jspurlin replied:
Can you reply with the bug info for the future work?

kelseyyoung replied:
https://github.com/OfficeDev/office-ui-fabric-react/issues/4800

joschect commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -485,58 +481,33 @@ export class ContextualMenu extends BaseComponent<IContextualMenuProps, IContext
 
   private _renderAnchorMenuItem(item: IContextualMenuItem, classNames: IMenuItemClassNames, index: number, focusableElementIndex: number, totalItemCount: number, hasCheckmarks: boolean, hasIcons: boolean): React.ReactNode {
+    const { contextualMenuItemAs } = this.props;
+    const dismissMenu = this.dismiss.bind(this, undefined);
```
Don't bind this here, you don't need to.

kelseyyoung replied:
I think I have to here. dismiss takes in the event as the first argument but I only pass through the boolean dismissAll in my callback. So I'm binding undefined to only pass through the second argument, basically

joschect replied:
The event can be undefined, it's fine to have it be undefined all the way up

joschect commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuAnchor.tsx:
```diff
@@ -0,0 +1,136 @@
+      </div>);
+  }
+
+  private _onItemMouseEnter = (ev: React.MouseEvent<HTMLElement>): void => {
```
Since these are the same as the ButtonItem methods, they should either get passed in as props, or inherited from some base component to reduce code duplication.

kelseyyoung commented on packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuAnchor.tsx:
```diff
@@ -0,0 +1,136 @@
+                openSubMenu={ openSubMenu }
+                dismissSubMenu={ dismissSubMenu }
+                dismissMenu={ dismissMenu }
+                getContainerElement={ this._getContainerElement }
```
will rename to 'getSubmenuTarget'

kelseyyoung commented:
Updated the PR, these are the changes:
Made a ContextualMenuItemWrapper base class. All three wrapper classes inherit from it. The shared functions are now protected in the base class. SplitButton still has a lot of private functions because of the large difference in it's functionality
